akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala

```diff
@@ -110,24 +129,124 @@ object Receptionist extends ExtensionId[Receptionist] {
+      }
+
+      def getServiceInstance[M](key: ServiceKey[M]): ActorRef[M] =
+        serviceInstance(key)
+    }
+
+    final case class Find[T] private[akka] (key: ServiceKey[T], replyTo: ActorRef[Receptionist.Listing]) extends Command
+
+    final case class Listing[T] private[akka] (key: ServiceKey[T], _serviceInstances: Set[ActorRef[T]]) extends Receptionist.Listing {
+
+      def isForKey(key: ServiceKey[_]): Boolean = key == this.key
+
+      def serviceInstances[M](key: ServiceKey[M]): Set[ActorRef[M]] = {
+        if (key != this.key) throw new IllegalArgumentException(s"Wrong key [$key] used, must use listing key [${this.key}]")
+        _serviceInstances.asInstanceOf[Set[ActorRef[M]]]
+      }
+
+      def getServiceInstances[M](key: ServiceKey[M]): java.util.Set[ActorRef[M]] =
+        serviceInstances(key).asJava
+    }
+
+    final case class Subscribe[T] private[akka] (key: ServiceKey[T], subscriber: ActorRef[Receptionist.Listing]) extends Command
+
+  }
+
   /**
    * Associate the given [[akka.actor.typed.ActorRef]] with the given [[ServiceKey]]. Multiple
    * registrations can be made for the same key. De-registration is implied by
    * the end of the referenced Actor’s lifecycle.
    *
-   * Registration will be acknowledged with the [[Registered]] message to the given replyTo actor.
+   * Registration will be acknowledged with the [[Registered]] message to the given replyTo actor
+   * if there is one.
    */
-  final case class Register[T](key: ServiceKey[T], serviceInstance: ActorRef[T], replyTo: ActorRef[Registered[T]]) extends Command
   object Register {
-    /** Auxiliary constructor to be used with the ask pattern */
-    def apply[T](key: ServiceKey[T], service: ActorRef[T]): ActorRef[Registered[T]] ⇒ Register[T] =
-      replyTo ⇒ Register(key, service, replyTo)
+
+    /**
+     * Create a Register without Ack that the service was registered
+     */
+    def apply[T](key: ServiceKey[T], service: ActorRef[T]): Command =
+      new MessageImpls.Register[T](key, service, None)
+    /**
+     * Create a Register with an actor that will get an ack that the service was registered
+     */
+    def apply[T](key: ServiceKey[T], service: ActorRef[T], replyTo: ActorRef[Registered]): Command =
+      new MessageImpls.Register[T](key, service, Some(replyTo))
   }
+  /**
+   * Java API: A Register message without Ack that the service was registered
+   */
+  def register[T](key: ServiceKey[T], service: ActorRef[T]): Command = Register(key, service)
+  /**
+   * Java API: A Register message with Ack that the service was registered
+   */
+  def register[T](key: ServiceKey[T], service: ActorRef[T], replyTo: ActorRef[Registered]): Command =
+    Register(key, service, replyTo)
 
   /**
    * Confirmation that the given [[akka.actor.typed.ActorRef]] has been associated with the [[ServiceKey]].
+   *
+   * Can not be pattern matched from Scala, use [[ServiceKey.Registered]] instead
```

Comment:
Maybe just say "You can use `key.Registered` for type-safe pattern matching.", that sounds more positive ;)

Reply:
I fixed this